async_fuse/src/memfs/dir.rs

```diff
@@ -1,46 +1,97 @@
 //! The implementation of directory related functionalities
 
-use nix::sys::stat::SFlag;
-use std::ffi::{CStr, OsStr, OsString};
+#[cfg(not(all(target_os = "linux", target_pointer_width = "64")))]
+compile_error!("async-fuse does not support this target now");
+
+use crate::util::{clear_errno, cstr_to_bytes, errno, with_c_str};
+
+use std::ffi::{OsStr, OsString};
+use std::io;
+use std::iter::FusedIterator;
 use std::os::unix::ffi::OsStrExt;
 use std::os::unix::io::RawFd;
-use std::ptr;
+use std::path::Path;
+use std::ptr::NonNull;
 
-use crate::fuse::protocol::INum;
+use libc::ino_t;
+use memchr::memchr;
+use nix::sys::stat::SFlag;
 
-#[cfg(target_os = "linux")]
-use libc::{dirent64 as dirent, readdir64_r as readdir_r};
+/// Directory meta-data
+pub struct Dir(NonNull<libc::DIR>);
 
-#[cfg(target_os = "macos")]
-use libc::{dirent, readdir_r};
+unsafe impl Send for Dir {}
+unsafe impl Sync for Dir {}
 
-/// Directory meta-data
-#[derive(Debug)]
-pub struct Dir(ptr::NonNull<libc::DIR>);
+impl Drop for Dir {
+    fn drop(&mut self) {
+        let dirp = self.0.as_ptr();
+        unsafe {
+            let _ = libc::closedir(dirp);
+        }
+    }
+}
 
 impl Dir {
-    /// Converts from a file descriptor, closing it on success or failure.
-    pub fn from_fd(fd: RawFd) -> nix::Result<Self> {
-        let d = unsafe { libc::fdopendir(fd) };
-        ptr::NonNull::new(d).map_or_else(
-            || {
-                let e = nix::Error::last();
-                unsafe { libc::close(fd) };
-                Err(e)
-            },
-            |non_null_ptr| Ok(Self(non_null_ptr)),
-        )
+    /// Calls opendir(3)
+    /// # Errors
+    /// This method will return `io::Error` if the underlying syscalls fails.
+    #[allow(dead_code)]
+    pub fn opendir<P: AsRef<Path>>(path: P) -> io::Result<Self> {
+        let bytes = path.as_ref().as_os_str().as_bytes();
+        with_c_str(bytes, |p| unsafe {
+            let dirname = p.as_ptr();
+            let dirp = libc::opendir(dirname);
+            if dirp.is_null() {
+                return Err(io::Error::last_os_error());
+            }
+            Ok(Self(NonNull::new_unchecked(dirp)))
```

Comment:
As the dirp is not null after the preview if check, shall we call new directly instead of new_unchecked?

Reply:
[NonNull::new](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.new) performs a null check which is equivalent to `if-else`.
`if dirp.is_null()` is more explicit.